src/Symfony/Component/Mailer/Bridge/Amazon/SesRequest.php

```diff
@@ -0,0 +1,245 @@
+
+        if ($emails = $email->getCc()) {
+            $payload['Destination.CcAddresses.member'] = $this->stringifyAddresses($emails);
+        }
+        if ($emails = $email->getBcc()) {
+            $payload['Destination.BccAddresses.member'] = $this->stringifyAddresses($emails);
+        }
+        if ($email->getTextBody()) {
+            $payload['Message.Body.Text.Data'] = $email->getTextBody();
+        }
+        if ($email->getHtmlBody()) {
+            $payload['Message.Body.Html.Data'] = $email->getHtmlBody();
+        }
+
+        $this->payload = $payload;
+
+        $this->prepareRequestHeaders();
+
+        return $this->sendRequest();
+    }
+
+    public function sendRawEmail(string $rawEmail): ResponseInterface
+    {
+        $this->now = new \DateTime();
+        $this->action = 'SendRawEmail';
+        $this->method = 'POST';
+        $this->payload = [
+            'Action' => 'SendRawEmail',
+            'RawMessage.Data' => base64_encode($rawEmail),
+        ];
+
+        $this->prepareRequestHeaders();
+
+        return $this->sendRequest();
+    }
+
+    private function sendRequest(): ResponseInterface
+    {
+        $options = [
+            'headers' => $this->requestHeaders,
+            'body' => $this->payload,
+        ];
+
+        return $this->client->request($this->method, 'https://'.$this->getEndpointHost(), $options);
+    }
+
+    private function prepareRequestHeaders(): void
+    {
+        $this->requestHeaders = [];
+
+        if (self::REQUEST_MODE_API === $this->mode) {
+            $this->requestHeaders['Content-Type'] = 'application/x-www-form-urlencoded';
+        }
+
+        $this->requestHeaders['Host'] = $this->getEndpointHost();
+        $this->requestHeaders['X-Amz-Date'] = $this->now->format('Ymd\THis\Z');
+        if ($this->credential instanceof ApiTokenCredential) {
+            $this->requestHeaders['X-Amz-Security-Token'] = $this->credential->getToken();
+        }
+        ksort($this->requestHeaders);
+
+        $canonicalHeadersBuffer = [];
+        foreach ($this->requestHeaders as $key => $value) {
+            $canonicalHeadersBuffer[] = strtolower($key).':'.$value;
+        }
+        $canonicalHeaders = implode("\n", $canonicalHeadersBuffer);
+
+        $signedHeadersBuffer = [];
+        foreach ($this->requestHeaders as $key => $value) {
+            $signedHeadersBuffer[] = strtolower($key);
+        }
+        $signedHeaders = implode(';', $signedHeadersBuffer);
+
+        $canonicalRequestBuffer = [];
+        $canonicalRequestBuffer[] = $this->method;
+        $canonicalRequestBuffer[] = '/';
+        $canonicalRequestBuffer[] = '';
+        $canonicalRequestBuffer[] = $canonicalHeaders;
+        $canonicalRequestBuffer[] = '';
+        $canonicalRequestBuffer[] = $signedHeaders;
+        $canonicalRequestBuffer[] = hash('sha256', $this->arrayToSignableString($this->payload));
+        $hashedCanonicalRequest = hash('sha256', implode("\n", $canonicalRequestBuffer));
```

Comment:
Imho and I might be nitpicking but, write this as:

```php
$hashedCanonicalRequest = hash('sha256', sprintf(
    '%s\n/\n\n%s\n\n%s\n%s', 
    $this->method, 
    $canonicalHeaders, 
    $signedHeaders, 
    hash('sha256', $this->arrayToSignableString($this->payload)))
);
```

Or like:

```php
$hashedCanonicalRequest = hash('sha256', implode("\n", [
    $this->method,
    '/',
    '',
    $canonicalHeaders,
    '',
    $signedHeaders,
    hash('sha256', $this->arrayToSignableString($this->payload))
]));
```

Same goes for the `$stringToSignBuffer` part below.

Reply:
I chose the sprintf because it's faster. thx